File: 0942. DI String Match/Solution.py
from typing import List


class Solution:
    def diStringMatch(self, S: str) -> List[int]:
        l, r, i = 0, len(S), 0
        ret = [0] * (r + 1)
        for ch in S:
            if ch == 'I':
                ret[i] = l
                l += 1
            else:
                ret[i] = r
                r -= 1
            i += 1
        ret[i] = l
        return ret

File: 0942. DI String Match/Solution.c
#include <stdlib.h>
#include <string.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *diStringMatch(char *S, int *returnSize)
{
    int N = strlen(S);
    int *ret = (int*)malloc(sizeof(int) * (N + 1));
    *returnSize = 0;
    int l = 0, r = N;
    for (int i = 0; i < N; i++)
        ret[(*returnSize)++] = S[i] == 'I' ? l++ : r--;
    ret[(*returnSize)++] = l;
    return ret;
}
